[ENH] 2d/3d Array Improvement

ryanhammonds opened the pull request:
**2d Array**
This addresses a point in #76 and is built on top of #75 (i.e. starting from deabc11). It allows for 2d arrays to be flatten prior to computing features using `axis = None`. The resulting dataframe is split into a list of dfs corresponding to indices in the original 2d signal. This is to address edge artifacts and features miscalculations from ignoring the influence of other signals.

For example flattening 2d to 1d (`axis=None`), or treating each signal separately (`axis=0`) will result in different burst fractions (`compute_burst_fraction`):
```
df_shape_features['volt_amp'].rank() / len(df_shape_features)
```

 - ``axis=1`` : Iterates over each row/signal in an array independently (i.e. for each channel
   in (n_channels, n_timepoints)).
 - ``axis=None`` : Flattens rows/signals prior to computing features (i.e. across flatten epochs
   in (n_epochs, n_timepoints)).

Another note, is that when  `axis = None`, the 2d to 1d array conversion prevents parallelization.

**3d Array**
An axis kwarg may be passed in `compute_features_3d`:

 - ``axis=0`` : Iterates over 2D slices along the zeroth dimension, (i.e. for each channel in
   (n_channels, n_epochs, n_timepoints)).
 - ``axis=1`` : Iterates over 2D slices along the first dimension (i.e. across flatten epochs in
   (n_epochs, n_timepoints)).
 - ``axis=(0, 1)`` : Iterates over 1D slices along the zeroth and first dimension (i.e across
   each signal independently in (n_participants, n_channels, n_timepoints)).

ryanhammonds commented:
This should be ready for a review. I've made a tutorial for how/when to use these functions in #74 and it has been pushed to the docsite [here](https://bycycle-tools.github.io/bycycle/v1.0.0rc/auto_examples/plot_2_multidim_sigs.html#sphx-glr-auto-examples-plot-2-multidim-sigs-py). Hopefully the tutorial makes sense! I would start with the tutorial before reviewing the code.

ryanhammonds commented:
Maybe add 2D/3D plotting funcs that stack 1D plots? MNE has nice functionality that lets you plot epochs vertically with interactive annotation / zooming features too - something I've looked into adding to ndspflow.

TomDonoghue commented:
Ahhh, sorry @ryanhammonds - I seemed to have missed to tag to review this one! I'll get to it soon!

TomDonoghue commented on bycycle/features/features.py:
```diff
@@ -83,13 +83,11 @@ def compute_features(sig, fs, f_range, center_extrema='peak', burst_method='cycl
-    df_samples : pandas.DataFrame, optional, default: True
-        An optionally returned dataframe containing cyclepoints for each cycle.
-        Columns (listed for peak-centered cycles):
+        When cyclepoints are returned (i.e. deafault, return_samples = True)
```
typo on `deafault`

TomDonoghue replied:
also - in comments do / should we write `return_samples = True` or `return_samples=True`. 
It looks a little weird to me to write this differently in comment than we would in code.

TomDonoghue commented on bycycle/features/features.py:
```diff
@@ -150,7 +146,8 @@ def compute_features(sig, fs, f_range, center_extrema='peak', burst_method='cycl
 
-    if return_samples:
-        return df_features, df_samples
+    if return_samples is False:
```
could / should this be a dataframe util? Just like we have 'split_samples_df', it could be `drop_samples_df` or something. I think it would be useful to have the utility if there may be some cases to do this operation outside this function.

TomDonoghue commented on bycycle/features/burst.py:
```diff
@@ -9,16 +9,13 @@
 
     Parameters
     ----------
     df_shape_features : pandas.DataFrame
```
doesn't this input now combine both shape & cyclepoint features? If so, this description seems perhaps slightly out of data. Maybe it should be called `df_features`?

TomDonoghue commented on bycycle/group/features.py:
```diff
@@ -74,91 +91,154 @@ def compute_features_2d(sigs, fs, f_range, compute_features_kwargs=None,
+        df_flat = compute_features(sig_flat, fs=fs, f_range=f_range, return_samples=True,
+                                   center_extrema=center_extrema, **kwargs[0])
+
+        # Reshape the dataframe into original sigs shape
```
so, this feels a bit like it's an independent, and potentially useful task: to 'epoch' a dataframe of measured values. 

Could this be a sub-function? As well as increasing modularity, I feel like it could be useful be itself. For example, if I've already computed features on a long continuous signal, I can imagine wanting to split it up into time sections. 

Is there a relatively easy way to turn this code section into something like:
`epoch_df(df, epoch_lengths)`
which would take a single df, and split it up into a list of dfs, each representing `epoch_length` samples?
This function could be used here, and offered as a general utility, I think?

ryanhammonds replied:
I refactored this to a new func  in `utils.dataframes`.

TomDonoghue commented on bycycle/group/features.py:
```diff
@@ -189,81 +269,90 @@ def compute_features_3d(sigs, fs, f_range, compute_features_kwargs=None,
+    df_features = np.zeros((np.shape(sigs)[0], np.shape(sigs)[1])).tolist()
 
-        if kwargs.ndim == 1 and np.shape(kwargs)[0] != np.shape(sigs)[0]:
+    if axis == 0:
```
The process for axis 0 or 1 is effectively the same, with just a bit of data manipulation depending which actual axis. 

This can therefore potentially be condensed, I think:

```
if axis in [0, 1]:
    sigs = np.swapaxes(sigs, 0, 1) is axis == 1 else sigs
    # then have the lines 287 to 295, that are otherwise duplicated in 300 to 308
    if axis == 1: df_features = [list(dfs) for dfs in zip(*df_features)]
```

ryanhammonds replied:
Good idea! Updated to this

TomDonoghue commented on bycycle/group/features.py:
```diff
@@ -74,91 +91,154 @@ def compute_features_2d(sigs, fs, f_range, compute_features_kwargs=None,
+        - ``axis = 1`` (default) : Features are computed for each signal independently
+          across the second axis, i.e. across channels in (n_epochs, n_channels, n_timepoints).
+        - ``axis = None`` : Features are computed across a flattened 1d array, this is typically
+          not recommended unless the 3d array was recored continuously.
```
Ohhh, wait. it's only when I finished reading the code that I realized this flattens -> 2D, and not all the way to 1D. 
in that case, this option feels equally sensible to the flatten option within `compute_features_2d`. I'm not sure either are super general, but I think they can plausibly be kept. But - this doc should be clear as to how it flattens.

ryanhammonds replied:
The 3D array does eventually get flatten to 1D (within `compute_features_2d`), when `axis=None`. The 2d func is used in this case to maintain a 2D dataframe structure that can more easily be reorganized into a 3D list of dataframes. I added this option to make the function as flexible as possible for all use cases. But now that you mention it, I don't think this would ever be useful and can probably be removed.

TomDonoghue commented on bycycle/group/features.py:
```diff
@@ -189,81 +269,90 @@ def compute_features_3d(sigs, fs, f_range, compute_features_kwargs=None,
+    sigs, kwargs = args[0], args[1]
 
-    return df_reshape
+    return compute_features_2d(sigs, fs=fs, f_range=f_range, return_samples=return_samples,
```
very much a nitpick, but for consistency, I would say order of arguments here should also match the function def.

TomDonoghue commented:
okay @ryanhammonds - I did and finished the deep dive review here. I think overall it looks nice, and has a nice element of being able to distribute the core functions across dimensions, allowing for lots of flexibility, without too much code overall. 

So, general approach is good, as usual I've left a bunch of specific comments and suggestions along the way! :)

ryanhammonds commented on bycycle/group/features.py:
```diff
@@ -74,91 +91,154 @@ def compute_features_2d(sigs, fs, f_range, compute_features_kwargs=None,
+          across the second axis, i.e. across channels in (n_epochs, n_channels, n_timepoints).
+        - ``axis = None`` : Features are computed across a flattened 1d array, this is typically
+          not recommended unless the 3d array was recored continuously.
+        - ``axis = (0, 1)`` : Features are computed independently for each signal, across the first
```
(0, 1) or tuples in general are supported by numpy. However, I realized axis=(0, 1) functions differently in np than it does here. In np , (0, 1) means to run across axis 0, take the resulting array and run across axis 1. I think it should be updated to axis=2 to match numpy.

```
In [1]: import numpy as np                                                                                                                   

In [2]: arr = np.ones((2, 3, 4))                                                                                                             

In [3]: arr                                                                                                                                  
Out[3]: 
array([[[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]],

       [[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]]])

In [4]: np.sum(arr, axis=0)                                                                                                                  
Out[4]: 
array([[2., 2., 2., 2.],
       [2., 2., 2., 2.],
       [2., 2., 2., 2.]])

In [5]: np.sum(arr, axis=1)                                                                                                                  
Out[5]: 
array([[3., 3., 3., 3.],
       [3., 3., 3., 3.]])

In [6]: np.sum(arr, axis=(0, 1))                                                                                                             
Out[6]: array([6., 6., 6., 6.])

In [7]: np.sum(arr, axis=2)                                                                                                                  
Out[7]: 
array([[4., 4., 4.],
       [4., 4., 4.]])
```

ryanhammonds commented on bycycle/group/features.py:
```diff
@@ -189,81 +269,90 @@ def compute_features_3d(sigs, fs, f_range, compute_features_kwargs=None,
-    # Ensure compute_features corresponds to sigs
-    if isinstance(kwargs, np.ndarray):
+    # Compute features
+    df_features = np.zeros((np.shape(sigs)[0], np.shape(sigs)[1])).tolist()
```
Good catch, thanks!

I renamed `df_features` to `dfs_features`.

ryanhammonds commented on tutorials/plot_3_bycycle_resting_state.py:
```diff
@@ -116,8 +116,8 @@
 
-plot_burst_detect_summary(df_features_list[1], df_samples_list[1], sigs[1], fs,
-                          threshold_kwargs, xlim=(0, 5), figsize=(16, 3))
+plot_burst_detect_summary(df_features_list[1], sigs[0], fs, threshold_kwargs,
```
It was a typo, good catch!

ryanhammonds commented on bycycle/utils/dataframes.py:
```diff
@@ -102,19 +103,22 @@ def get_extrema_df(df):
         Which extrema is centered.
-    df_samples, df_features : pandas.DataFrames
-        ByCycle dataframes to rename, given the centered extrema.
+    df_shape_features : pandas.DataFrames
```
Yep, thanks!

ryanhammonds commented:
@TomDonoghue I addressed each of your comments. Thanks for the feedback!

I think the biggest change in the last few commits was changing how the axis arg works. It functions the exact same as the numpy axis arg now. I replaced `axis=(0, 1)`  with `axis=2` for 3D array. `axis=0` was also replaced with `axis=1` for 2D arrays. This was to functionally match the numpy axis arg.

This should be ready to merge, but I'm going to finish updating the tutorials before merging.

TomDonoghue commented:
@ryanhammonds - sounds good - I had a quick scroll, and all looks good to me!
